### PR TITLE
Add compiler macro foreign-type-size

### LINKED ISSUE #340: Expanding foreign-type-size for standard types
On a fresh SBCL instance, the function `foreign-type-size` is called, even for constant expressions eval'd to standard types:

```lisp
* (require :cffi)
NIL
* (defun foo ()
    (declare (optimize speed))
    (cffi:foreign-type-size :int))
FOO
* (disassemble #'foo)
; disassembly for FOO
; Size: 22 bytes. Origin: #x5366769B                          ; FOO
; 9B:       BA5F873F50       MOV EDX, #x503F875F              ; ':INT
; A0:       B902000000       MOV ECX, 2
; A5:       FF7508           PUSH QWORD PTR [RBP+8]
; A8:       B8629C3F50       MOV EAX, #x503F9C62              ; #<FDEFN CFFI:FOREIGN-TYPE-SIZE>
; AD:       FFE0             JMP RAX
; AF:       CC10             INT3 16                          ; Invalid argument count trap
```

While a custom type may be redefined to be a different size, I don't think there ought to be a reason why types like `:int`, `:uint` etc should change, especially for implementations whose compiled output aren't portable across architectures.

Here's a first pass what I think a compiler macro for this could look like:

```lisp
(define-compiler-macro foreign-type-size (&whole whole foreign-type
                                                 &environment env)
    (flet ((standard-type-p (type)
             (or (keywordp type)
                 (and (lisp type) (eq (first type) :pointer)))))
      (let ((expanded-type (macroexpand foreign-type env))
            evaled-type)
        (if (and (constantp expanded-type env)
                 (standard-type-p (setf evaled-type (eval expanded-type)))
                 (ignore-errors (foreign-type-size evaled-type)))
            (foreign-type-size evaled-type)
            whole))))
```

### PR DESCRIPTION
**Pull request by samuel-hunter**
Fixes #340. Depends on #345 

The compiler macro is a first-pass that should fix most use cases without breakage. It checks to see if the foreign type is constant, and if so, expands into its foreign type size, falling back to the old form if an error is signaled.

Before the compiler macro:

```lisp
* (disassemble '(lambda () (cffi:foreign-type-size '(:pointer :uint32))))
; disassembly for (LAMBDA ())
; Size: 32 bytes. Origin: #x5365FC5C                          ; (LAMBDA ())
; 5C:       498B4510         MOV RAX, [R13+16]                ; thread.binding-stack-pointer
; 60:       488945F8         MOV [RBP-8], RAX
; 64:       488B15BDFFFFFF   MOV RDX, [RIP-67]                ; '(:POINTER
                                                              ;   :UINT32)
; 6B:       B902000000       MOV ECX, 2
; 70:       FF7508           PUSH QWORD PTR [RBP+8]
; 73:       B8A29D3F50       MOV EAX, #x503F9DA2              ; #<FDEFN CFFI:FOREIGN-TYPE-SIZE>
; 78:       FFE0             JMP RAX
; 7A:       CC10             INT3 16                          ; Invalid argument count trap
NIL
```

After (on SBCL 2.2.5 64-bit)

```lisp
* (disassemble (lambda () (cffi:foreign-type-size '(:pointer :uint32))))
; disassembly for (LAMBDA ())
; Size: 21 bytes. Origin: #x5365FECC                          ; (LAMBDA ())
; CC:       498B4510         MOV RAX, [R13+16]                ; thread.binding-stack-pointer
; D0:       488945F8         MOV [RBP-8], RAX
; D4:       BA10000000       MOV EDX, 16
; D9:       488BE5           MOV RSP, RBP
; DC:       F8               CLC
; DD:       5D               POP RBP
; DE:       C3               RET
; DF:       CC10             INT3 16                          ; Invalid argument count trap
NIL
```

Tested on SBCL and CCL with no additional test errors.

### CODE REVIEW

File: src/early-types.lisp [hunk 1]
```diff
@@ -164,6 +164,14 @@ Signals an error if FOREIGN-TYPE is undefined."))
   (:documentation
    "Return the size in bytes of a foreign type."))
 
+(define-compiler-macro foreign-type-size (&whole form foreign-type)
+  ;; Don't pass &environment to CONSTANTP or MACROEXPAND as it may make
+  ;; surprising behavior with LOAD-TIME-VALUE.
+  (if-let ((comptime-value (when (constantp (macroexpand foreign-type))
+                             (ignore-errors (eval form)))))
```
Comment (phoe) [inline src/early-types.lisp:171]: `ignore-errors` + `eval` looks like a pretty big code smell to me; wouldn't `(locally (declare (notinline foreign-type-size)) (foreign-type-size foreign-type))` work better? It should ensure that the function is called at compilation time without recursively invoking the compiler macro again.

File: src/early-types.lisp [hunk 2]
```diff
@@ -164,6 +164,14 @@ Signals an error if FOREIGN-TYPE is undefined."))
   (:documentation
    "Return the size in bytes of a foreign type."))
 
+(define-compiler-macro foreign-type-size (&whole form foreign-type)
+  ;; Don't pass &environment to CONSTANTP or MACROEXPAND as it may make
+  ;; surprising behavior with LOAD-TIME-VALUE.
```
Comment (phoe) [inline src/early-types.lisp:169]: Could you provide more examples of this?

File: src/early-types.lisp [hunk 3]
```diff
@@ -164,6 +164,14 @@ Signals an error if FOREIGN-TYPE is undefined."))
   (:documentation
    "Return the size in bytes of a foreign type."))
 
+(define-compiler-macro foreign-type-size (&whole form foreign-type)
```
Comment (phoe) [inline src/early-types.lisp:167]: Are there any tests that should be performed for this compiler macro alone? I assume that you could call the compiler macro function yourself to ensure that it provides correct results on some test cases.

File: src/early-types.lisp (hunk 1 above)
Comment (samuel-hunter) [reply src/early-types.lisp:171]: Oh, nice! This was what I was looking for before I gave up on `(l-t-v ,form)` -> eval.

File: src/early-types.lisp (hunk 2 above)
Comment (samuel-hunter) [reply src/early-types.lisp]: I rephrased this now that I took a second look at this. Clarified: since L-T-V executes under a null lexical environment, we shouldn't be checking if the foreign type is constant under &environment.

**Comment by sionescu** [comment]
@samuel-hunter Copy QUOTEDP, CONSTANTP, and EVAL-CONSTANT from https://github.com/sionescu/static-vectors/blob/master/src/constantp.lisp and use them.
We'll figure out later how to share the code.

**Comment by samuel-hunter** [comment]
> @samuel-hunter Copy QUOTEDP, CONSTANTP, and EVAL-CONSTANT from https://github.com/sionescu/static-vectors/blob/master/src/constantp.lisp and use them. We'll figure out later how to share the code.

Preexisting compiler macros ought to benefit from this if grafted into CFFI. Should this be split into its own PR?

**Comment by sionescu** [comment]
Why a new PR ?

**Comment by samuel-hunter** [comment]
I'll make a new branch/PR at least for now

**Comment by sionescu** [comment]
@samuel-hunter Please rebase to master. I added the utils in d4216a3acafa.

**Comment by sionescu** [comment]
I opened https://github.com/cffi/cffi/issues/344 for converting existing compiler macros.

File: src/early-types.lisp [hunk 4]
```diff
@@ -164,6 +164,13 @@ Signals an error if FOREIGN-TYPE is undefined."))
...
+  ;; Don't pass &environment to CONSTANTP or MACROEXPAND as LOAD-TIME-VALUE is
+  ;; executed in a null lexical environment.
+  (if (constantp (macroexpand foreign-type))
+      `(load-time-value (locally (declare (notinline foreign-type-size)) ,form))
```
Comment (sionescu) [inline src/early-types.lisp:171]: It's better to use CONSTANT-FORM-VALUE here. L-T-V inhibits futher optimisations, such as (foreign-alloc (+ <constant> (foreign-type-size :foo)))

Comment (samuel-hunter) [reply src/early-types.lisp]: Fixed, thanks :+1:

File: src/early-types.lisp [hunk 5]
```diff
@@ -165,10 +165,8 @@ Signals an error if FOREIGN-TYPE is undefined."))
    "Return the size in bytes of a foreign type."))
 
 (define-compiler-macro foreign-type-size (&whole form foreign-type)
-  ;; Don't pass &environment to CONSTANTP or MACROEXPAND as LOAD-TIME-VALUE is
-  ;; executed in a null lexical environment.
-  (if (constantp (macroexpand foreign-type))
-      `(load-time-value (locally (declare (notinline foreign-type-size)) ,form))
+  (if (constant-form-p (macroexpand foreign-type))
```
Comment (sionescu) [inline src/early-types.lisp:168]: Let's do the conversion in another PR. You can cherry-pick this commit to another branch.

**Review by sionescu — CHANGES_REQUESTED** [review]

File: src/early-types.lisp (hunk 4 above)
Comment (sionescu) [reply src/early-types.lisp:171]: Did you undo the change to use CONSTANT-FORM-VALUE ?

**Review by sionescu — CHANGES_REQUESTED** [review]

File: src/early-types.lisp (hunk 4 above)
Comment (samuel-hunter) [reply src/early-types.lisp]: I don't *think* so...? Unless I did that in process of reordering the commits.
